lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,164 @@
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    ops::Deref,
+    path::PathBuf,
+};
+
+use sha2::{Digest, Sha256};
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`ModuleCache::load()`] will
+/// be called more often than [`ModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait ModuleCache: Debug {
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError>;
```

Comment:
Thinking more about this, probably this should return a `LoadCacheError` (enum with states: NotFound, ReadError, Generic), as is more specialized and `WriteError` and `SerializeError` are states that can't be reached here

Reply:
Let's leave it as a single error for now. There's a good chance the `ModuleCache` trait's shape will change as we start getting feedback from production, and if we add `purge()` or `remove()` methods I don't really want the `wasmer_wasix::runtime::module_cache` to keep growing extra error types.

Reply:
The question I ask myself with error types is: what invariants need to be handled with custom code paths ? 

Sometimes hard to tell before you use it. 

I think this is fine for now. 
(the main one I can think of is `deserialize`)